### PR TITLE
Modules type

### PR DESCRIPTION
**Pull request by dvdherron**
## Description
Upgrading the `type` module  with Sass module usage. 

## Steps to test/reproduce

See files in `./sass/type/`

_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

## Show me
_Provide screenshots/animated gifs/videos if necessary._

### CODE REVIEW

**Comment by dvdherron** [comment]
[I want Accoutrement on Sass Modules](https://trello.com/c/9tyINy2o/23-i-want-accoutrement-on-sass-modules)

**Comment by mirisuzanne** [comment]
[I want Accoutrement on Sass Modules](https://trello.com/c/9tyINy2o/23-i-want-accoutrement-on-sass-modules)

**Comment by dvdherron** [comment]
@mirisuzanne I'm working through updating this module and fixing the tests. 

Tests are passing for `config`, `a11y`, and `pseudo-elements`.  I still need to work on `font-vars`, `fonts` and `helpers`.

Three tests are failing in `normalize`. (I've commented them out).  Can you tell if there's something wrong in the tests or in `type/normalize` that's causing them to fail?

File: sass/type/_normalize.scss [hunk 1]
```diff
@@ -129,14 +129,6 @@
...
   $name: $name or inspect.is-alias-for($source, $key, 'origin') or $key;
 
-  @if (meta.type-of($name) == 'string') {
-    $name: if(
-      string.index($name, ''),
-      string.quote($name),
-      string.unquote($name)
-    );
-  }
-
```
Comment (mirisuzanne) [inline sass/type/_normalize.scss]: It may have been a mistake to remove this… but at the time it seemed redundant, and didn't have any tests.

File: test/type/_normalize.scss [hunk 2]
```diff
@@ -6,25 +6,26 @@
 @use '../../../sass/tokens/api';
 @use '../../../sass/type/normalize';
 
-// Test Font Normalization
-// =======================
+$_global: config.$fonts;
+config.$fonts: data.$fonts;
```
Comment (mirisuzanne) [inline test/type/_normalize.scss:10]: All these functions take a `$source` argument. If it's not set explicitly, they look up `config.$fonts`. We need to make sure they find the actual source map we're trying to test.

File: test/type/_normalize.scss [hunk 3]
```diff
@@ -6,25 +6,26 @@
...
       @include assert-equal(map.get($test, 'name'), 'TestFont');
-    } */
+    }
 
-    /*     @include it('Generates a properly-formatted stack') {
+    @include it('Generates a properly-formatted stack') {
       @include assert-equal(
         map.get($test, 'stack'),
-        ('TestFont', Georgia, serif)
+        (TestFont, Georgia, serif)
```
Comment (mirisuzanne) [inline test/type/_normalize.scss:25]: The quoting and unquoting of names in a font stack is important. Output should only have quotes when the font name in question has spaces.

File: test/type/_normalize.scss (hunk 2 above)
Comment (dvdherron) [reply test/type/_normalize.scss:10]: Ah, ok. I saw you doing this in the test for `colors` and now that makes sense.

File: sass/type/_normalize.scss (hunk 1 above)
Comment (dvdherron) [reply sass/type/_normalize.scss]: hmm. without these lines, the first two tests in `Fonts` fail due to the extra quotes:
image

Comment (mirisuzanne) [reply sass/type/_normalize.scss]: super weird that they didn't fail for me. but good catch - @dvdherron can you add these lines back in? (since this is used a couple places, it could also become a function if you wanted)

Comment (dvdherron) [reply sass/type/_normalize.scss]: yeah, strange. And I will look into turning this into a function . . .

File: sass/type/_helpers.scss [hunk 4]
```diff
@@ -147,19 +144,15 @@ $_a_FONT-FORMATS: (
...
-) {
-  $format-name: map-get($_a_FONT-FORMATS, $format);
+@function _a_font-src($path, $format) {
+  $format-name: map.get($FONT-FORMATS, $format);
   $format-name: if($format-name, format($format-name), '');
 
   // stylelint-disable-next-line function-url-quotes
-  $path: _a_font-url($path, $format, $iefix: true);
+  $path: font-url($path, $format, $iefix: true);
   @return font-url($path) $format-name;
```
Comment (dvdherron) [inline sass/type/_helpers.scss:153]: @mirisuzanne As you know, the second `font-url` function here is defined in the `fonts` module. if I use that module here and try to call it with `fonts.font-url` I get a module loop error:

```
Error: Module loop: this module is already being loaded.
  ┌──> sass/type/_helpers.scss
5 │ @use './fonts';
  │ ^^^^^^^^^^^^^^ new load
  ╵
  ┌──> test/type/_fonts.scss
3 │ @use '../../../sass/type/fonts';
  │ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ original load
```
Would the best way to go about this be to relocate/rename the `font-url` function in `fonts`?

Comment (mirisuzanne) [reply sass/type/_helpers.scss:153]: For sure we should rename one or both of the font-url functions, to clarify what they do. But then, yeah, we can't have both modules relying on the other - so _something_ has to relocate. Maybe it works to move the `fonts` one into helpers? Since that's the public one, it's also the one that should get priority for keeping the name.

Comment (dvdherron) [reply sass/type/_helpers.scss:156]: @mirisuzanne 
I ended up moving the `fonts` version into `helpers` and renamed the `helpers` version to `format-url`.

All `type` tests are passing now except for one in `fonts`.

 ```
Import Webfonts [mixin]
         Font-face imports for all local fonts & variants:

```

If I remove the two FranklinGothic `@font-face` blocks, the test passes. I think this means that the test is looking for those two fonts in the global `config.$fonts map`. 

When I add FranklinGothic to the `test-data` font map, it makes the first test in type/config fail
image

I think that tests assumes that Franklin isn't in the global fonts map, right? 

Should I just remove to two `@font-face` blocks for `FranklinGothic` in the `Import Webfonts [mixin]` test? 

We can schedule a time to walk through this if this is hard to follow :)

File: sass/type/_font-vars.scss [hunk 5]
```diff
@@ -25,12 +31,11 @@
 ///    )
 /// );
 /// $font-var-prefix: 'prefix_';
```
Comment (dvdherron) [inline sass/type/_font-vars.scss:33]:
```suggestion
/// tools.$font-var-prefix: 'prefix_';
```

File: sass/type/_font-vars.scss [hunk 6]
```diff
@@ -25,12 +31,11 @@
...
-/// html { @include font--('my-font') }
+/// html { @include tools.font--('my-font') }
 ///
 /// @type string
 $font-var-prefix: 'font-' !default;
```
Comment (dvdherron) [inline sass/type/_font-vars.scss:37]:
```suggestion
tools.$font-var-prefix: 'font-' !default;
```

File: sass/type/_index.scss [hunk 7]
```diff
@@ -7,35 +7,23 @@
...
-@import 'helpers';
-@import 'normalize';
-@import 'a11y';
-@import 'fonts';
-@import 'font-vars';
-@import 'pseudo-elements';
-
+@forward 'config';
+@forward 'helpers';
+@forward 'normalize';
```
Comment (dvdherron) [inline sass/type/_index.scss:12]:
```suggestion

```

File: sass/type/_font-vars.scss [hunk 8]
```diff
@@ -192,22 +191,29 @@ $font-var-prefix: 'font-' !default;
...
-) {
+@function font-values($source: config.$fonts, $value: 'stack') {
   $return: ();
 
   @each $key, $raw in $source {
-    $normal: get-token($source, $key);
-    $normal: _a_normalize-font($normal, $key, $source);
-    $has: map-get($normal, $value);
+    $normal: api.get($source, $key);
+    $normal: normalize.normalize-font($normal, $key, $source);
```
Comment (dvdherron) [inline sass/type/_font-vars.scss:199]:
```suggestion
    $normal: normalize.font($normal, $key, $source);
```

File: sass/type/_fonts.scss [hunk 9]
```diff
@@ -30,45 +38,12 @@
...
 ///   A parsed and normalized map of font-data
-@function font(
-  $font,
-  $source: $fonts
-) {
-  $data: get-token($source, $font);
-  $normal: _a_normalize-font($data, $font, $source);
+@function font($font, $source: config.$fonts) {
+  $data: api.get($source, $font);
+  $normal: normalize.normalize-font($data, $font, $source);
```
Comment (dvdherron) [inline sass/type/_fonts.scss:43]:
```suggestion
  $normal: normalize.font($data, $font, $source);
```

File: sass/type/_fonts.scss [hunk 10]
```diff
@@ -168,35 +135,38 @@
...
-  $source: $fonts
-) {
+@mixin font-face($key, $source: config.$fonts) {
   $font: font($key, $source);
-  $variants: _a_font-get-variants($font);
-  $name: map-get($font, 'name');
+  $variants: helpers.font-get-variants($font);
+  $name: map.get($font, 'name');
 
   @if (length($variants) > 0) {
```
Comment (dvdherron) [inline sass/type/_fonts.scss:151]:
```suggestion
  @if (list.length($variants) > 0) {
```

File: sass/type/_index.scss [hunk 11]
```diff
@@ -7,35 +7,23 @@
...
-@import 'config';
-@import 'helpers';
-@import 'normalize';
-@import 'a11y';
-@import 'fonts';
-@import 'font-vars';
-@import 'pseudo-elements';
-
+@forward 'config';
+@forward 'helpers';
```
Comment (dvdherron) [inline sass/type/_index.scss:11]:
```suggestion

```

File: sass/type/_helpers.scss [hunk 12]
```diff
@@ -363,8 +361,36 @@ $_a_FONT-FORMATS: (
...
       font-display: $display;
       src: $eot;
-      src: _a_font-src-list($data);
+      src: font-src-list($data);
       unicode-range: $unicode;
     }
   }
 }
+
+// Font-URL [function]
```
Comment (dvdherron) [inline sass/type/_helpers.scss:370]: public, so should be forwarded. use @forward show ...

File: sass/type/_helpers.scss [hunk 13]
```diff
@@ -363,8 +361,36 @@ $_a_FONT-FORMATS: (
...
+// -------------------
+/// This is a wrapper for the CSS `url()` function,
+/// which simply prepends the global `$font-path` when appropriate.
+///
+/// @since type-4.0.0 -
+/// - BUGFIX: Supports Base64 data URI's, without prepending font-path
+///
+/// @group type-fonts
+/// @example scss
+///   $font-path: '../fonts/';
```
Comment (dvdherron) [inline sass/type/_helpers.scss:380]:
```suggestion
///   tools.$font-path: '../fonts/';
```

File: sass/type/_helpers.scss [hunk 14]
```diff
@@ -363,8 +361,36 @@ $_a_FONT-FORMATS: (
...
+///   $src: font-url('sans/my-sans-font.woff2');
```
Comment (dvdherron) [inline sass/type/_helpers.scss:381]:
```suggestion
///   $src: tools.font-url('sans/my-sans-font.woff2');
```

File: sass/type/_normalize.scss [hunk 15]
```diff
@@ -24,42 +34,59 @@
 ///   to use as the origin palette
 /// @return {map} -
 ///   A map of font data in consistent longhand format
-@function _a_normalize-font(
-  $font,
-  $key: null,
-  $source: $fonts
-) {
+@function normalize-font($font, $key: null, $source: config.$fonts) {
```
Comment (dvdherron) [inline sass/type/_normalize.scss:37]:
```suggestion
@function normalize.font($font, $key: null, $source: config.$fonts) {
```

File: sass/type/_normalize.scss [hunk 16]
```diff
@@ -24,42 +34,59 @@
...
+  @if (meta.type-of($font) != 'map') {
+    $font: (
+      'stack': $font,
+    );
   }
 
   // Normalize name…
-  $name: _a_normalize-font-name($font, $key, $source);
-  $font: if($name, map-merge($font, ('name': $name)), $font);
+  $name: normalize-font-name($font, $key, $source);
```
Comment (dvdherron) [inline sass/type/_normalize.scss:46]:
```suggestion
  $name: font-name($font, $key, $source);
```

File: sass/type/_normalize.scss [hunk 17]
```diff
@@ -24,42 +34,59 @@
...
+        'name': $name,
+      )
+    ),
+    $font
+  );
 
   // Normalize the font stack…
-  $stack: _a_normalize-font-stack($font);
-  $font: if($stack, map-merge($font, ('stack': $stack)), $font);
+  $stack: normalize-font-stack($font);
```
Comment (dvdherron) [inline sass/type/_normalize.scss:59]:
```suggestion
  $stack: font-stack($font);
```

**Comment by dvdherron** [comment]
@mirisuzanne @stacyk 

I cleaned up based on the review:
- made `normalize` and `helper` private
- renamed the functions in normalize so that they would not be double referenced. 
- and fixed the missing module references 

It seems like I'm missing loading a module somewhere though. When I remove the type files from the `IGNORE_SASS` list in the gulpfile, I get errors about undefined mixins and variables in the examples:
```
» [WARNING] Error compiling @example scss: 
Undefined mixin.
   ╷
3  │ ┌ @include tools.import-font-face(
4  │ │   $name: 'Franklin Gothic',
5  │ │   $weight: 'normal',
6  │ │   $style: 'normal',
7  │ │   $data: (
8  │ │     'local': 'Franklin Gothic' ('franklin-gothic'),
9  │ │     'woff2': 'path/to/franklin-bolditalic-font.woff2',
10 │ │     'ttf': 'path/to/franklin-bolditalic-font.ttf',
11 │ │   )
12 │ │ );
```
   │ └─^

I'll take a closer look at what's happening here tomorrow.

File: sass/type/_font-vars.scss (hunk 6 above)
Comment (mirisuzanne) [reply sass/type/_font-vars.scss:37]: This is the initial definition of the variable, so should not be prefixed.

File: sass/type/_index.scss [hunk 18]
```diff
@@ -68,22 +57,22 @@
 ///
 /// ```scss
 ///  // Import one font by configuration key
-///  @include font-face('body');
+///  @include tools.font-face('body');
```
Comment (mirisuzanne) [inline sass/type/_index.scss:60]: Don't prefix markdown examples? (see multiple on this page)

**Review by mirisuzanne — COMMENTED** [review]
Looking pretty good, just needs some final cleanup?

**Review by mirisuzanne — APPROVED** [review]
